harness/tests/experiment/pytorch/test_deepspeed_autotuning.py

```diff
@@ -312,39 +362,348 @@ def test_trial_queue_and_state_all_errors(self, basic_trial_tracker):
             metrics = list(reversed(metrics))
         while trial_tracker.can_run_more_trials:
-            popped_trial = trial_tracker.pop_next_trial()
+            popped_trial = trial_tracker.queue.popleft()
             trial_tracker.update_trial_metric(
                 popped_trial, {popped_trial.searcher_metric_name: metrics.pop()}
             )
             assert trial_tracker.best_trial == popped_trial
             assert trial_tracker.best_trials_by_stage[popped_trial.stage] == popped_trial
 
 
+def random_search_builder(args):
+    """
+    Creates a `RandomDSATSearchMethod` instance with a completed model profile info run.
+    """
+    exp_config = get_custom_dsat_exp_conf_from_args(args)
+    search_method = _dsat_search_method.RandomDSATSearchMethod(args=args, exp_config=exp_config)
+    searcher_state = _search_method.SearcherState()
+    search_method.initial_operations(searcher_state)
+    search_method.on_validation_completed(
+        searcher_state,
+        search_method.trial_tracker.model_profile_info_trial.request_id,
+        MODEL_INFO_PROFILE_METRIC_FIXTURE,
+        search_method.trial_tracker.model_profile_info_trial.length,
+    )
+    return searcher_state, search_method
+
+
+@pytest.fixture
+def default_random_search_method():
+    searcher_state, search_method = random_search_builder(DEFAULT_ARGS)
+    yield searcher_state, search_method
+
+
+class TestRandomDSATSearchMethodTrialCreation:
+    """
+    Testing the various `RandomDSATSearchMethod` methods related to trial creation.
+    """
+
+    def test_random_hparams_and_search_data(self, default_random_search_method):
+        _, search_method = default_random_search_method
+        for _ in range(100):
+            for stage in range(4):
+                hparams, search_data = search_method.get_random_hparams_and_search_data(stage)
+                mbs = hparams[_defaults.OVERWRITE_KEY]["train_micro_batch_size_per_gpu"]
+                assert hparams[_defaults.OVERWRITE_KEY]["zero_optimization"]["stage"] == stage
+                assert search_data.lo <= mbs <= search_data.hi
+
+    def test_random_hparams_and_search_data_after_best(self, default_random_search_method):
+        for _ in range(100):
+            _, search_method = default_random_search_method
+            for stage in range(4):
+                hparams, search_data = search_method.get_random_hparams_and_search_data(stage)
+                trial = search_method.trial_tracker.create_trial(hparams, search_data)
+                search_method.trial_tracker.queue_and_register_trial(trial)
+                search_method.trial_tracker.queue.popleft()
+                search_method.trial_tracker.update_trial_metric(
+                    trial, {trial.searcher_metric_name: 0.0}
+                )
+                _, new_search_data = search_method.get_random_hparams_and_search_data(stage)
+                assert new_search_data.lo <= new_search_data.hi
```

Comment:
I would recommend adding some similar tests to explicitly set the hi and lo equal to each other and make sure that it holds the constraints properly